src/hotspot/share/utilities/unsigned5.hpp

```diff
@@ -0,0 +1,419 @@
+    const uint32_t b_0 = (uint8_t) get(array, pos);  //b_0 = a[0]
+    assert(b_0 >= X, "avoid excluded bytes");
+    uint32_t sum = b_0 - X;
+    if (sum < L) {  // common case
+      offset_rw = pos + 1;
+      return sum;
+    }
+    // must collect more bytes:  b[1]...b[4]
+    int lg_H_i = lg_H;  // lg(H)*i == lg(H^^i)
+    for (int i = 1; ; i++) {  // for i in [1..4]
+      assert(limit == 0 || pos + i < limit, "oob");
+      const uint32_t b_i = (uint8_t) get(array, pos + i);  //b_i = a[i]
+      assert(b_i >= X, "avoid excluded bytes");
+      sum += (b_i - X) << lg_H_i;  // sum += (b[i]-X)*(64^^i)
+      if (b_i < X+L || i == MAX_LENGTH-1) {
+        offset_rw = pos + i + 1;
+        return sum;
+      }
+      lg_H_i += lg_H;
+    }
+  }
+
+  // encode a single unsigned 32-bit int into an array-like span
+  // offset_rw is both read and written
+  // the limit is either zero meaning no limit check, or an exclusive offset
+  // warning:  caller must ensure there is available space
+  template<typename ARR, typename OFF, typename SET = ArrayGetSet<ARR,OFF>>
+  static void write_uint(uint32_t value, ARR array, OFF& offset_rw, OFF limit, SET set = SET()) {
+    const OFF pos = offset_rw;
+    if (value < L) {
+      const uint32_t b_0 = X + value;
+      assert(b_0 == (uint8_t)b_0, "valid byte");
+      set(array, pos, (uint8_t)b_0);  //a[0] = b_0
+      offset_rw = pos + 1;
+      return;
+    }
+    uint32_t sum = value;
+    for (int i = 0; ; i++) {  // for i in [0..4]
+      if (sum < L || i == MAX_LENGTH-1) {
+        // remainder is either a "low code" or the 5th byte
+        uint32_t b_i = X + sum;
+        assert(b_i == (uint8_t)b_i, "valid byte");
+        set(array, pos + i, (uint8_t)b_i);  //a[i] = b_i
+        offset_rw = pos + i + 1;
+        return;
+      }
+      sum -= L;
+      uint32_t b_i = X + L + (sum % H);  // this is a "high code"
+      assert(b_i == (uint8_t)b_i, "valid byte");
+      set(array, pos + i, (uint8_t)b_i);  //a[i] = b_i
+      sum >>= lg_H;                 // extracted 6 bits
+    }
+  }
+
+  // returns the encoded byte length of an unsigned 32-bit int
+  static constexpr int encoded_length(uint32_t value) {
+    // model the reading of [0..5] high-bytes, followed possibly by a low-byte
+    // Be careful:  the constexpr magic evaporates if undefined behavior
+    // results from any of these expressions.  Beware of signed overflow!
+    uint32_t sum = 0;
+    uint32_t lg_H_i = 0;
+    for (uint32_t i = 0; ; i++) {  // for i in [1..4]
+      if (value <= sum + ((L-1) << lg_H_i) || i == MAX_LENGTH-1) {
+        return i + 1;  // stopping at byte i implies length is i+1
+      }
+      sum += (MAX_b - X) << lg_H_i;
+      lg_H_i += lg_H;
+    }
+  }
+
+  // reports the largest uint32_t value that can be encoded using len bytes
+  // len must be in the range [1..5]
+  static constexpr uint32_t max_encoded_in_length(uint32_t len) {
```

Comment:
A few of the classes make the gtest test a friend so that internal functions can be private.